File: Koko Eating Bananas - GFG/koko-eating-bananas.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution {
  public:
    int Solve(int N, vector<int>& piles, int h) {
        int low = 1;
        int high = *max_element(piles.begin(),piles.end());

        while(low <= high){
            int k = (low+high) >> 1;

            long long int hour = 0;
            for(auto i : piles){
                if(i <= k)
                    hour++;
                else{
                    hour += (i/k);
                    if(i % k)
                        hour++;
                }
            }

            if(hour > h)
                low = k+1;
            else
                high = k-1;
        }

        return low;
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        int n;
        cin >> n;
        vector<int> a;
        for (int i = 0; i < n; ++i) {
            int x;
            cin >> x;
            a.push_back(x);
        }
        int h;
        cin >> h;
        Solution obj;

        cout << obj.Solve(n, a, h);
        cout << "\n";
    }
    return 0;
}
// } Driver Code Ends

File: 907-koko-eating-bananas/koko-eating-bananas.py
class Solution:
    def eating_hour(self, piles, k):
        h = 0

        for i in piles:
            h += (i // k)

            if i % k != 0:
                h += 1
        
        return h

    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        n = len(piles)
        low = 1
        high = max(piles)

        while low <= high:
            mid = (low + high) // 2

            est_h = self.eating_hour(piles, mid)

            if est_h <= h:
                high = mid - 1
            else:
                low = mid + 1
        
        return low